### PR TITLE
use untyped yaml configuration as function or connector config

### PR DESCRIPTION
**Pull request by freeznet**
#231

### CODE REVIEW

**Review by nlu90 — APPROVED** [review]

**Comment by freeznet** [comment]
@tuteng @fantapsody PTAL when you have time, thanks.

**Review by tuteng — APPROVED** [review]

File: controllers/spec/common.go [hunk 1]
```diff
@@ -423,7 +423,10 @@ func generateResource(resources corev1.ResourceList) *proto.Resources {
 	}
 }
 
-func getUserConfig(configs map[string]string) string {
+func getUserConfig(configs *v1alpha1.Config) string {
+	if configs == nil {
+		return ""
```
Comment (fantapsody) [inline controllers/spec/common.go:428]: "" might not be a valid json, will that be a problem?

File: api/v1alpha1/common.go [hunk 2]
```diff
@@ -251,3 +253,49 @@ const (
...
+		return err
+	}
+	c.Data = out
+	return nil
+}
+
+// DeepCopyInto is an ~autogenerated~ deepcopy function, copying the receiver, writing into out. in must be non-nil.
+// This exists here to work around https://github.com/kubernetes/code-generator/issues/50
+func (c *Config) DeepCopyInto(out *Config) {
+	bytes, err := json.Marshal(c.Data)
```
Comment (fantapsody) [inline api/v1alpha1/common.go:288]: Could https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/converter.go#L437 be more efficient?

**Review by fantapsody — APPROVED** [review]

File: controllers/spec/common.go (hunk 1 above)
Comment (freeznet) [reply controllers/spec/common.go:428]: the `userConfig` in pulsar functions' proto `FunctionDetails` is defined as string, and the default value is `""`, so `""` would be fine if there is no `userConfig` provided.

**Pull request merged**